### ISSUE TITLE
Memory competition between notebook processes - proactively run GC?

### ISSUE THREAD

**Issue by BioTurboNick** [fonsp/Pluto.jl#1850]
I have an issue where there's a shared Pluto instance and multiple people are using notebooks. They leave them open, and in total the notebooks are absorbing a huge chunk of system memory.

I've noticed that Pluto memory usage can grow really large to fill available memory, without running the GC. We're talking 4-5 GB when after GC it's 2.5 GB, in one instance.

I think it would also help to have Pluto run the GC on a notebook process automatically to reduce its footprint on the system. Either when no cells are pending, or some time after the notebook has been idle?

**Comment by dralletje** [comment]
Can you create a MVP example notebook?
Like one where you don't GC at the end of cells and we can see the memory go up, and one where you put a forced GC at the end of cells and we can see the memory be lower?

**Comment by BioTurboNick** [comment]
@fonsp - Just wanted to bump this.

It seems like it would be easy enough to implement? After all cells have run, execute `GC.gc()`, and possibly `ccall(:malloc_trim, Int32, (Int32,), 0)`

I'd do a PR but not sure where to put it.

**Comment by fonsp** [comment]
Hi @BioTurboNick , thanks!

My only concern is making interactive notebooks slower, e.g. `@bind x Slider(1:100)` and `plot(rand(x))`. It looks like the overhead of calling `GC.gc()` is at least 50ms, which is too much to allow for this case.

Perhaps we could run it only after the initial run, or only after a manual run (Shift+Enter), not a bond change.

What is `ccall(:malloc_trim, Int32, (Int32,), 0)`? Do you have any links to learn more? It's best if we don't use internal Julia API unless absolutely necessary.

**Comment by BioTurboNick** [comment]
Fair point about adjusting sliders... could there be a cancelable task spawned that waits some duration and is reset each time cells are run?

Re: malloc_trim: https://github.com/JuliaLang/julia/issues/42566#issuecomment-1319476638 (the process in question was a Pluto notebook)

Basically, Julia frees memory with the GC but doesn't always release it back to the OS. This is good if Julia needs to allocate that memory again soon because it doesn't have to make a syscall. This may be improved in a later release.

**Comment by BioTurboNick** [comment]
Doing it on first run would be reasonable; though making it on command isn't ideal because it depends on users knowing they should do it.

**Comment by BioTurboNick** [comment]
Just to add that a common pattern I think is to put long/intensive executions behind checkboxes or button. It would be nice if this feature could run then too, somehow. Like if the execution was longer than 1 s, the overhead of the GC may be acceptable.

**Comment by fonsp** [comment]
Can GC run safely in a separate thread?

**Comment by fonsp** [comment]
RE: malloc_trim: I think it's best to wait a bit for https://github.com/JuliaLang/julia/issues/42566 to progress. Adding `ccall(:malloc_trim, Int32, (Int32,), 0)` could potentially create segfaults if the current Pluto version is used in a future Julia version where the API is removed (right?). That means that we also can't `try catch` it for future compatibility. 

But adding a `GC.gc()` call would already be a nice improvement!

**Comment by pankgeorg** [comment]
> Could be generalized into some kind of "idle" state in which we can execute background tasks

Yes; you also don't want the GC to starve, as it may be stressing the execution.

**Comment by BioTurboNick** [comment]
@fonsp - I'm interested in trying a PR on this. Could you please point me to where a hook exists, or could be placed, to trigger such a mechanism?

And if you also know of similar places to hook into user actions on a notebook (loading an open notebook into a tab, performing some other interaction), that would be useful for #2236

**Comment by pankgeorg** [comment]
A kind-of educated guess could be to run the GC step before (or using) `NotebookExecutionDoneEvent`; i.e: here https://github.com/fonsp/Pluto.jl/blob/4d9596bfa6ce01559154687da76fa5945608c9b9/src/evaluation/Run.jl#L34 and here https://github.com/fonsp/Pluto.jl/blob/4d9596bfa6ce01559154687da76fa5945608c9b9/src/evaluation/Run.jl#L452-L455

You don't need a PR to test this, you can try

```julia
import Pluto

function onevent(ev::Pluto.PlutoEvent) end
function onevent(ev::Pluto.NotebookExecutionDoneEvent)
    condition = true
	condition && GC.gc()
end
otherOptions = Dict{Symbol, Any}(#=Add more here =#)
Pluto.run(;on_event=onevent, otherOptions...)
```
